experimental/ast/decl_body.go

```diff
@@ -68,17 +63,37 @@ func (d DeclScope) Span() Span {
 }
 
 // Len returns the number of declarations inside of this body.
-func (d DeclScope) Len() int {
+func (d DeclBody) Len() int {
 	return len(d.raw.ptrs)
 }
 
 // At returns the nth element of this body.
-func (d DeclScope) At(n int) Decl {
-	return d.raw.kinds[n].reify().with(d.Context(), d.raw.ptrs[n])
+func (d DeclBody) At(n int) DeclAny {
+	k := d.raw.kinds[n]
+	p := d.raw.ptrs[n]
+
+	switch k {
+	case DeclKindEmpty:
+		return wrapDeclEmpty(d, arena.Pointer[rawDeclEmpty](p)).AsAny()
+	case DeclKindSyntax:
+		return wrapDeclSyntax(d, arena.Pointer[rawDeclSyntax](p)).AsAny()
+	case DeclKindPackage:
+		return wrapDeclPackage(d, arena.Pointer[rawDeclPackage](p)).AsAny()
+	case DeclKindImport:
+		return wrapDeclImport(d, arena.Pointer[rawDeclImport](p)).AsAny()
+	case DeclKindDef:
+		return wrapDeclDef(d, arena.Pointer[rawDeclDef](p)).AsAny()
+	case DeclKindBody:
+		return wrapDeclBody(d, arena.Pointer[rawDeclBody](p)).AsAny()
+	case DeclKindRange:
+		return wrapDeclRange(d, arena.Pointer[rawDeclRange](p)).AsAny()
+	default:
+		panic(fmt.Sprintf("protocompile/ast: unknown declKind %d: this is a bug in protocompile", k))
+	}
```

Comment:
Feels a bit circuitous and verbose to use a switch to create the right concrete type and then immediately go back to the generic form. You could instead trivially construct the generic result with a one-liner:
```suggestion
	return DeclAny{
		withContext: d.withContext,
		ptr:         d.raw.ptrs[n],
		kind:        d.raw.kinds[n],
	}
```

Reply:
Slick!